jerry-core/lit/lit-strings.c

```diff
@@ -316,6 +316,48 @@ lit_get_utf8_size_of_cesu8_string (const lit_utf8_byte_t *cesu8_buf_p, /**< cesu
 } /* lit_get_utf8_size_of_cesu8_string */
 
 /**
+ * Calculate length of an utf-8 encoded string from cesu-8 encoded string
+ *
+ * @return length of an utf-8 encoded string
+ */
+ecma_length_t
+lit_get_utf8_length_of_cesu8_string (const lit_utf8_byte_t *cesu8_buf_p, /**< cesu-8 string */
+                                     lit_utf8_size_t cesu8_buf_size) /**< string size */
+{
+  lit_utf8_size_t offset = 0;
+  ecma_length_t utf8_length = 0;
+
+  while (offset < cesu8_buf_size)
+  {
+    ecma_char_t ch;
+    offset += lit_read_code_unit_from_utf8 (cesu8_buf_p + offset, &ch);
+    utf8_length++;
+
+    if (lit_is_code_point_utf16_high_surrogate (ch) && (offset < cesu8_buf_size))
+    {
+      ecma_char_t next_ch;
+      lit_utf8_size_t next_ch_size = lit_read_code_unit_from_utf8 (cesu8_buf_p + offset, &next_ch);
+
+      if (lit_is_code_point_utf16_high_surrogate (next_ch))
+      {
+        continue;
+      }
```

Comment:
Is there a way to optimize this without decoding the character twice?

Perhaps we could "remember" the previous character, and if we found a low surrogate, we check that the previous character was a high surrogate.